internal/ingest/ingest.go

```diff
@@ -943,30 +941,21 @@ func (ing *Ingester) RunWorkers(n int) {
 	for n > ing.workerPoolSize {
 		// Start worker.
 		ing.waitForWorkers.Add(1)
-		go ing.ingestWorker(ing.workersCtx)
+		go ing.ingestWorker(ing.workersCtx, ing.syncFinishedEvents)
 		ing.workerPoolSize++
 	}
 	for n < ing.workerPoolSize {
 		// Stop worker.
-		ing.closeWorkers <- struct{}{}
+		ing.stopWorker <- struct{}{}
 		ing.workerPoolSize--
 	}
 }
 
-// syncWatcher handles events from dagsync.Subscriber signaling that an
-// advertisement chain has been synced. These events are passed to the
-// chainQueue for processing by a wokrer.
-func (ing *Ingester) syncWatcher(syncFinishedEvents <-chan dagsync.SyncFinished) {
-	for event := range syncFinishedEvents {
-		ing.chainQueue <- event
-	}
-	close(ing.syncWatcherDone)
-}
-
-// ingestWorker processes events from the chainQueue to create provider work
-// assignments from a raw advertisement chain, and processes those provider
-// work assignments.
-func (ing *Ingester) ingestWorker(ctx context.Context) {
+// ingestWorker processes events from dagsync.Subscriber, signaling that an
+// advertisement chain has been synced. Provider work assignments are created
+// from raw advertisement chains, and processed by ingestWorker. Work
+// assignments are processed preferentially over new advertisement chains.
+func (ing *Ingester) ingestWorker(ctx context.Context, syncFinishedEvents <-chan dagsync.SyncFinished) {
```

Comment:
Note: `syncFinishedEvents` is passed in as an argument so that if a test needs to replace `ing.syncFinishedEvents` , it does not have to wait for all the workers to start and then stop first.